### ISSUE TITLE
[Tracking] Handling multiple versions of CUDA libraries in the same closure

### ISSUE THREAD

**Issue by ConnorBaker** [NixOS/nixpkgs#341650]
### Describe the bug

When multiple versions of CUDA libraries are in the same closure, the order they are discovered and loaded by different packages is unclear.

This is issue is meant to track progress toward documentation, tools, and tests to communicate and ensure that libraries are loaded in a predictable and deterministic fashion.

- https://github.com/NixOS/nixpkgs/pull/341471

#### Related Links

- https://github.com/NixOS/nixpkgs/pull/339619

### Notify maintainers

@NixOS/cuda-maintainers

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by samuela** [comment]
This seems like a Hard Problem, and I'm not aware of any package managers that actually solve it. Is there a need to fix this?

**Comment by Princemachiavelli** [comment]
When I need to manually debug these kinds of issues, I use `LD_DEBUG=libs` but it's output isn't easy to parse. You might find using the `LD_AUDIT` [interface](https://linux.die.net/man/7/rtld-audit) easier and more informative. Unfortunately the only user space tool that implements it, `latrace`, appears to be unmaintained and not functional. Luckily the `LD_AUDIT` interface is only a few functions.

Is the thought that CUDA libraries are loading in a non-predictable fashion due to the use of `dlopen` or some source of non-determinism effecting the dynamic linker `/lib/ld-linux.so.2`?

**Comment by samuela** [comment]
> Is the thought that CUDA libraries are loading in a non-predictable fashion due to the use of dlopen or some source of non-determinism effecting the dynamic linker /lib/ld-linux.so.2?

That's my understanding. Other uses should be explicit in RUNPATH/RPATH.

**Comment by SomeoneSerge** [comment]
> You might find using the LD_AUDIT [interface](https://linux.die.net/man/7/rtld-audit) easier and more informative

Oh! We were thinking of parsing LD_DEBUG but you're making a good point! FWIW there are more examples of using LD_AUDIT, e.g. in `guix pack`, `flox`, and now https://github.com/cachix/devenv/issues/773#issuecomment-2239772997

**Comment by SomeoneSerge** [comment]
However, note that for this particular test we don't even have to interact with ld.so in any particular way, we just verify that the program does not crash when using two potentially conflicting modules

**Comment by hacscred** [comment]
I also see you opened CUDA issues in 2023. Do you not have any intention to solve them? 

I don't think this problem is a "Hard Problem". 

I would expect a large bug bounty for such an issue, if it were important. Normal users don't generally have multiple different generation of CUDA in one program. It's so specialized that whoever wants it could easily pay for its development and such a person would benefit from it being maintained in the future.

**Comment by SomeoneSerge** [comment]
> Normal users don't generally have multiple different generation of CUDA in one program

This is unfortunately more common than one might expect: opencv, pytorch, tensorflow, jax, all have their own fairly narrow and conflicting compatibility constraints wrt CUDA and in general, yet users often have to use several of those in the same environment.

We try to relax their constraints and enforce fixpoints where possible, but on many occasions we do have to mix versions. In these cases the only reasonable thing we could do is run a CI to obtain empirical evidence that mixing doesn't cause problems

> I don't think this problem is a "Hard Problem".

How'd you approach this?

**Comment by hacscred** [comment]
A tracking issue only makes sense if there's someone that will actually work on it eventually. 

I was interested in learning about why so little progress has been made. 

I think anyone improving the value of a $3T for free is irrational. I would like AI hardware, but if it's not sufficiently open, my expertise isn't free. It only devalues the profession to do things that mostly benefit large companies for free. Anyone eventually doing it, should just be handsomely paid for it. 

Why are *you* "tracking" this? 

In general, combining different library versions is actually impossible depending on the specific hardware and programs involved. It might work in some cases, as you claim. 

It would be pointless to explain an approach, because anyone that would understand the explanation should also be able to think of the solution. It took me all of 10 seconds to figure out a solution that would work for everyone. The details are probably a month or two of work. 

The economic value of having it work for everyone (not just on NixOS) would probably be hundreds of millions of dollars. That's not bad for a few months of work. I am not seeing a Kickstarter yet, though.

**Comment by SomeoneSerge** [comment]
@hacscred your manner of communication is inappropriate. If it's not obvios why, and if you've really written any of this text and not generated it using LLMs, we can have a chat over voice/video to avoid further misunderstandings

**Comment by RuRo** [comment]
> hacscred your manner of communication is inappropriate.

Agreed. @SomeoneSerge and @ConnorBaker are major contributors to nixpkgs. They can have whatever tracking issues that they want. And they don't owe anyone their work.

I wanted to take this opportunity to tell you guys that you really are doing God's work in nixpkgs and I hope that you aren't discouraged by such negative interactions. ❤️

**Comment by hacscred** [comment]
@SomeoneSerge What specifically is inappropriate and who made you the judge of that? I think you should write more code and less fluff. If that means you write no code and no "tracking issues" I am sure the rest of humanity would be sad. Also, implying that I am a LLM isn't totally inappropriate? How about you actually get some perspective on human relations? Why would I ever be interested in talking to you face to face? There would be nothing of value you could tell me. 

You seem to have great difficulty in figuring out when someone has a misunderstanding. I am not misunderstanding anything. You seem to have a misunderstanding, which is not of interest to me, since most people misunderstand most of the things in the world. Stop harassing me. I was already long done in this thread. 

@RuRo When did I say they owed anyone work? I was basically saying they were littering the bug tracker by adding issues they apparently don't solve and nobody apparently cares about. Would it be possible to not accuse me of things I didn't do? That's slander. Also, I didn't even say anything about ConnorBaker (nor am I such an annoying person to ping them for nothing, like you apparently like to do). In fact, Serge doesn't accept my position, which implies he wants me to provide $100,000+ in value for free for merely explaining a superior solution. 

If you are someone's greatest fan, start a fan club, but leave me out of it. 

And finally, for both of you: There's nothing reasonable about anything anyone of you have said.
